### PR TITLE
Receptor docs

### PR DESCRIPTION
**Pull request by fosterseth**
currently published here https://fosterseth-receptor.readthedocs.io/en/latest/index.html

Covers basic receptor concepts and usage. It does not explain every parameter possible in `receptor --help` (like `port`), rather focuses on the more obscure ones, or ones that are used very commonly.

### CODE REVIEW

File: docs/read-the-docs/source/index.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+- `receptorctl`: A user-facing command line used to interact with receptor, typically over a Unix domain socket.
+- `node`: A single running instance of receptor.
+- `node ID`: An arbitrary string identifying a single node, analogous to an IP address.
+- `backend`: A type of connection that receptor nodes can pass traffic over. Current backends include TCP, UDP and websockets.
+- `control service`: A built-in service that usually runs under the name `control`.  Used to report status and to launch and monitor work.
+
+Installation
+^^^^^^^^^^^^
+
+Compile from source code (golang 1.15 required)
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/index.rst]: `1.15.x`

File: docs/read-the-docs/source/index.rst [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+- `receptor`: The receptor application taken as a whole, that typically runs as a daemon.
+- `receptorctl`: A user-facing command line used to interact with receptor, typically over a Unix domain socket.
+- `node`: A single running instance of receptor.
+- `node ID`: An arbitrary string identifying a single node, analogous to an IP address.
+- `backend`: A type of connection that receptor nodes can pass traffic over. Current backends include TCP, UDP and websockets.
+- `control service`: A built-in service that usually runs under the name `control`.  Used to report status and to launch and monitor work.
+
+Installation
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/index.rst]: Consider breaking into `Building` vs. `Installing`. Regardless of if you break it up, the install part is kind of missing. Maybe add "Copy the receptor binary into something in your `$PATH`" @shanemcd may have a better recommended install.

Should the recommended install method be build the rpm then yum install it?

File: docs/read-the-docs/source/index.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+receptorctl will be used in various places throughout this documentation.
+
+Basic usage
+^^^^^^^^^^^
+
+Run the following command in a terminal to start a node called `foo`
+
+.. code-block::
+
+    receptor --node id=foo --local-only --log-level Debug
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/index.rst]: I'm like a cave man with a stick. I like to be able to poke things and see how they respond. Could this example be modified to be pokable and then poked (i.e. ping example)?

Maybe align with the container default config and add a control service.

```
- control-service:
    service: control
    filename: /tmp/receptor.sock
```

Then we can run `receptorctl status`

File: docs/read-the-docs/source/index.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+Container image
+^^^^^^^^^^^^^^^
+
+.. code-block::
+
+    podman pull quay.io/project-receptor/receptor
+
+The default receptor configuration file in the container is located at ``/etc/receptor/receptor.conf``
+
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/index.rst]: Maybe add an example run and a poke.

```
podman run -it --rm --name receptor quay.io/project-receptor/receptor

INFO2021/08/04 18:41:03 Running control service control
INFO2021/08/04 18:41:03 Initialization complete
```

```
podman exec receptor /usr/local/bin/receptorctl status

Node ID: 9b612286e561
Version: 1.0.0
System CPU Count: 8
System Memory MiB: 31491

Node         Service   Type       Last Seen           Tags            Work Types
9b612286e561 control   Stream     2021-08-04 18:45:12 -               -
```

File: docs/read-the-docs/source/connecting_nodes.rst [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+bar.yml
+
+.. code-block:: yaml
+
+    ---
+    - node:
+        id: bar
+
+    -log-level:
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/connecting_nodes.rst]: missing a space `-log-level` -> `- log-level`

File: docs/read-the-docs/source/connecting_nodes.rst [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+.. image:: mesh.png
+
+foo.yml
+
+.. code-block:: yaml
+
+    ---
+    - node:
+        id: foo
+
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/connecting_nodes.rst]: add a control-service so we can poke it.

```
- control-service:
    service: control
    filename: /tmp/receptor.sock
```

File: docs/read-the-docs/source/connecting_nodes.rst [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    INFO 2021/07/22 23:04:31 Known Connections:
+    INFO 2021/07/22 23:04:31    fish: foo(1.00)
+    INFO 2021/07/22 23:04:31    foo: bar(1.00) fish(1.00)
+    INFO 2021/07/22 23:04:31    bar: foo(1.00)
+    INFO 2021/07/22 23:04:31 Routing Table:
+    INFO 2021/07/22 23:04:31    foo via foo
+    INFO 2021/07/22 23:04:31    bar via foo
+
+
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/connecting_nodes.rst]: The state of the mesh can be gotten by asking the receptor node `foo`.

```
receptorctl --socket /tmp/receptor.sock status
Node ID: foo
Version:
System CPU Count: 8
System Memory MiB: 31491

Connection   Cost
fish         1
bar          1

Known Node   Known Connections
bar          {'foo': 1}
fish         {'foo': 1}
foo          {'bar': 1, 'fish': 1}

Route        Via
bar          bar
fish         fish

Node         Service   Type       Last Seen           Tags                            Work Types
foo          control   Stream     2021-08-04 15:11:38 -                               -
```

File: docs/read-the-docs/source/interacting_with_nodes.rst [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+bar.yml
+
+.. code-block:: yaml
+
+    ---
+    - node:
+        id: bar
+
+    -log-level:
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/interacting_with_nodes.rst]: missing a space `-log-level` -> `- log-level`

File: docs/read-the-docs/source/workceptor.rst [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+bar.yml
+
+.. code-block:: yaml
+
+    ---
+    - node:
+        id: bar
+
+    -log-level:
```
Comment (chrismeyersfsu) [inline docs/read-the-docs/source/workceptor.rst]: missing a space `-log-level` -> `- log-level`

**Review by chrismeyersfsu — COMMENTED** [review]
I read `Connecting nodes` and wrote comments that are exactly addressed with examples in the `Interacting with nodes` section. So ignore my comments/mark as resolved. I figure I would keep them in so you can see the stream of consciousness coming from a first-time reader.

File: docs/read-the-docs/source/index.rst (hunk 4 above)
Comment (fosterseth) [reply docs/read-the-docs/source/index.rst]: done

File: docs/read-the-docs/source/index.rst [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+.. code-block::
+
+    receptor --help
+    receptor --version
+
+The preferred way to interact with receptor nodes is to use the receptorctl command line tool
+
+.. code-block::
+
+    pip install receptorctl
```
Comment (jladdjr) [inline docs/read-the-docs/source/index.rst]: `pip install -e receptorctl` to ensure that the receptorctl version matches the version of receptor just created with `make receptor`?

File: docs/read-the-docs/source/connecting_nodes.rst [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+``cost``  User-defined metric that will be used by the mesh routing algorithm. If the mesh were represented by a graph node, then cost would be the length or weight of the edges between nodes. When the routing algorithm determines how to pass network packets from one node to another, it will use this cost to determine an efficient path.
+
+``nodecost`` Cost to a particular node on the mesh, and overrides whatever is set in ``cost``.
+
+.. code-block::
+
+    - tcp-listener:
+        port: 2222
+        cost: 1.0
+        nodecost:
```
Comment (jladdjr) [inline docs/read-the-docs/source/connecting_nodes.rst]: Can `nodecost` be used for `tcp-peer` as well? (Guessing not, might be good to clarify here)

File: docs/read-the-docs/source/interacting_with_nodes.rst [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    - log-level:
+        level: Debug
+
+    - tcp-peer:
+        address: localhost:2222
+
+    - control-service:
+        service: control
+
+If ``filename`` is set, receptor will create a unix domain socket. Use receptorctl to interact with the running receptor node via this domain socket (using "--socket").
```
Comment (jladdjr) [inline docs/read-the-docs/source/interacting_with_nodes.rst]: How would you use receptorctl to connect with node if `filename` wasn't set?

File: docs/read-the-docs/source/connecting_nodes.rst (hunk 11 above)
Comment (fosterseth) [reply docs/read-the-docs/source/connecting_nodes.rst]: okay you make me realize I was missing an essential piece -- the cost must match on both ends of the connection. So for example fish's tcp-peer must have a cost of 2.0, to match the 2.0 on the tcp-listener in foo. I updated this part of the docs

File: docs/read-the-docs/source/workceptor.rst [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        port: 2222
+
+    - control-service:
+        service: control
+        filename: /tmp/foo.sock
+
+    - work-command:
+        workType: echosleep
+        command: bash
+        params:  "-c \"for i in {1..5}; do echo $i; done\""
```
Comment (jladdjr) [inline docs/read-the-docs/source/workceptor.rst]: The `echosleep` command in the example `echo`s, but does not actually `sleep`

Comment (jladdjr) [reply docs/read-the-docs/source/workceptor.rst]: (`bar`'s `echosleep` does actually sleep)

Comment (fosterseth) [reply docs/read-the-docs/source/workceptor.rst]: fixed, thanks

File: docs/read-the-docs/source/workceptor.rst [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+
+Local work
+^^^^^^^^^^
+
+Start the work by connecting to the ``control-service`` and issuing a "work submit" command
+
+.. code-block::
+
+    $ receptorctl --socket /tmp/foo.sock work submit echosleep --no-payload
```
Comment (jladdjr) [inline docs/read-the-docs/source/workceptor.rst]: `work submit` command is missing target node argument (this was listed as required in the `interacting_with_nodes.rst` section)

File: docs/read-the-docs/source/workceptor.rst [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+    $ receptorctl --socket /tmp/foo.sock work submit echosleep --node bar --no-payload
+    Result:  Job Started
+    Unit ID: 87Vwqb6A
+
+Sequence of events for remote work submission
+
+- `foo` starts a local work unit of work type "remote". This is a special work type that is built into receptor.
+- This work unit attempts to connect to `bar`'s control service and issue a "work submit echosleep" command. Note, from `bar`'s perspective, this is the exact same operation as if a user connected to `bar` directly and issued a work submit command. `bar` is not aware that `foo` is the one that issued the command.
+- Once submitted, `foo` will stream work results back to `foo` and store it on disk. It also periodically gets the ``status`` of the work running on `bar`. Status includes information about the work state and the stdout size.
```
Comment (jladdjr) [inline docs/read-the-docs/source/workceptor.rst]: > Once submitted, `foo` will stream work results back to `foo`

`bar` will stream work results back to `foo`?

Comment (jladdjr) [reply docs/read-the-docs/source/workceptor.rst]: or is this saying that `foo` will request that work results be streamed back to itself?

File: docs/read-the-docs/source/workceptor.rst [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+.. code-block::
+
+    $ receptorctl --socket /tmp/foo.sock work submit echosleep --node bar --no-payload
+    Result:  Job Started
+    Unit ID: 87Vwqb6A
+
+Sequence of events for remote work submission
+
+- `foo` starts a local work unit of work type "remote". This is a special work type that is built into receptor.
```
Comment (jladdjr) [inline docs/read-the-docs/source/workceptor.rst]: It is probably worth calling out that the Unit ID for the `remote` work unit  will almost certainly differ from the Unit ID for the `echosleep` command. When searching for the result of work, I've gotten a bit mixed up about the two IDs.

Also, while I'm thinking about this.. is there any way to correlate a `remote` Unit ID to the corresponding Unit ID of the actual command carried out by another node?

File: docs/read-the-docs/source/workceptor.rst [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    - work-command:
+        worktype: echosleep
+        command: bash
+        params:  "-c \"for i in {1..10}; do echo $i; sleep 1; done\""
+
+
+Configuring work commands
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``worktype`` User-defined name to give this work definition
```
Comment (jladdjr) [inline docs/read-the-docs/source/workceptor.rst]: Are there any restrictions on the characters that can be used in a worktype name? Restricted to alphanumeric + underscores / dashes?

File: docs/read-the-docs/source/workceptor.rst (hunk 15 above)
Comment (fosterseth) [reply docs/read-the-docs/source/workceptor.rst]: yes the latter, maybe `foo will stream work results back to itself` is more clear

File: docs/read-the-docs/source/k8s.rst [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+      - kubernetes namespace to use
+      - allowruntimeauth
+    * - secret_kube_config
+      - kubeconfig to authenticate with
+      - allowruntimeauth
+    * - secret_kube_pod
+      - pod definition
+      - allowruntimepod
+    * - pod_pending_timeout
+      - allowed duration for pod to be Pending
```
Comment (jladdjr) [inline docs/read-the-docs/source/k8s.rst]: what are the units on this? seconds? minutes?

File: docs/read-the-docs/source/workceptor.rst (hunk 14 above)
Comment (fosterseth) [reply docs/read-the-docs/source/workceptor.rst]: yes, receptorctl does a lot of convenient things like defaulting to the local node if --node is not provided.

I think I need to make it more clear in that section in interacting_with_nodes that the required/optional parameters are for when connecting to the socket directly. Under the hood, receptorctl will be following the requirements on that table, but the user won't be aware of it.

File: docs/read-the-docs/source/index.rst [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+receptorctl will be used in various places throughout this documentation.
+
+Basic usage
+^^^^^^^^^^^
+
+Run the following command in a terminal to start a node called `foo`
+
+.. code-block::
+
+    receptor --node id=foo --local-only -- log-level Debug
```
Comment (jladdjr) [inline docs/read-the-docs/source/index.rst]: What other log-levels are supported?

File: docs/read-the-docs/source/tls.rst [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+    ---
+    - node:
+        id: bar
+
+    - log-level:
+        level: Debug
+
+    - tls-client:
+        name: myclient
+        rootcas: ca.crt
```
Comment (jladdjr) [inline docs/read-the-docs/source/tls.rst]: would this just look for the file in the local directory, or does receptor search canonical folders used to hold certs on the system?

**Review by jladdjr — APPROVED** [review]

**Comment by jladdjr** [comment]
@fosterseth I feel like I got a bit nit-picky the further along I got. Take what's useful and forget the rest!

Nice job! These docs are ✨ . Thanks for writing everything up!

File: docs/read-the-docs/source/index.rst [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+Introduction
+============
+
+Receptor is an overlay network intended to ease the distribution of work across a large and dispersed collection of workers. Receptor nodes establish peer-to-peer connections with each other via existing networks. Once connected, the receptor mesh provides datagram (UDP-like) and stream (TCP-like) capabilities to applications, as well as robust unit-of-work handling with resiliency against transient network failures.
+
+Terminology and Concepts
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+- `receptor`: The receptor application taken as a whole, that typically runs as a daemon.
```
Comment (beeankha) [inline docs/read-the-docs/source/index.rst]: nitpick: `which typically runs as a daemon` ("which" vs "that")

File: docs/read-the-docs/source/interacting_with_nodes.rst [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+
+Note: "localhost" is a special keyword that tells receptor to connect to its own control-service. "localhost" can be used in all other control service commands that expect a node ID.
+
+From here, one can issue commands like "status" or "work list". Any "work submit" command will require a payload. You can type out the payload contents and press ctrl-D to send the EOF signal. Note that this will close the socket.
+
+Control service commands
+^^^^^^^^^^^^^^^^^^^^^^^^
+A ``control-service`` can accept commands in two formats; a space-delimited string or JSON. In some cases, JSON accepts arguments that are not supported in the string format.
+
+string example
```
Comment (beeankha) [inline docs/read-the-docs/source/interacting_with_nodes.rst]: String example:

File: docs/read-the-docs/source/interacting_with_nodes.rst [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+
+``work submit bar echosleep``
+
+JSON example
```
Comment (beeankha) [inline docs/read-the-docs/source/interacting_with_nodes.rst]: Add colon at end of text here

File: docs/read-the-docs/source/k8s.rst [hunk 24]
```diff
@@ -0,0 +1,92 @@
+Kubernetes Work
+===============
+
+Workceptor has the ability to launch Kubernetes pods to perform work
```
Comment (beeankha) [inline docs/read-the-docs/source/k8s.rst]: Add period at end of sentence here.

File: docs/read-the-docs/source/tls.rst [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+        cert: foo.crt
+        key: foo.key
+        requireclientcert: true
+        clientcas: ca.crt
+
+    - tcp-listener:
+        port: 2222
+        tls: myserver
+
+Having ``tls-server`` defined does nothing on its own, but it can be referenced elsewhere in the receptor config file. Above, ``tls`` in the ``tcp-listener`` is set to use ``myserver``. In general, ``tls-server`` should be referenced any where receptor is expecting an incoming connection, i.e. ``*-listener`` backends or on the ``control-service``. Similarly, ``tls-client`` should be referenced any where receptor is expecting to make an outgoing connection, i.e. ``*-peer`` backends or in receptorctl (the command-line client for receptor).
```
Comment (beeankha) [inline docs/read-the-docs/source/tls.rst]: Change instances of `any where` to `anywhere`

**Review by beeankha — APPROVED** [review]
This is awesome!

File: docs/read-the-docs/source/developer_guide.rst [hunk 26]
```diff
@@ -0,0 +1,324 @@
+Developer guide
+================
+
+Receptor is an open source project that lives at https://github.com/ansible/receptor
+
+Testing
+^^^^^^^
+
+Pull requests must pass a suite of integration tests before being merged into devel.
```
Comment (beeankha) [inline docs/read-the-docs/source/developer_guide.rst]: Nitpick: Either make `devel` at the end of this sentence be "monospace", or re-write to say "Pull requests must pass a suite of integration tests before being merged into the default development branch"

File: docs/read-the-docs/source/developer_guide.rst [hunk 27]
```diff
@@ -0,0 +1,324 @@
...
+
+``make test`` will run the full test suite locally. Some of the tests require access to a Kubernetes cluster, and expects to load in the kubeconfig located at ``$HOME/.kube/config``. One simple way to make these tests work is to start minikube locally before running ``make test``. See https://minikube.sigs.k8s.io/docs/start/ for more information about minikube.
```
Comment (beeankha) [inline docs/read-the-docs/source/developer_guide.rst]: "Some of the tests require access to a Kubernetes cluster, and expects to load in the kubeconfig located at `$HOME/.kube/config`." 

Reword to:
"Some of the tests require access to a Kubernetes cluster; these tests will load in the `kubeconfig` file located at `$HOME/.kube/config`."

File: docs/read-the-docs/source/developer_guide.rst [hunk 28]
```diff
@@ -0,0 +1,324 @@
...
+
+To skip tests that depend on Kubernetes, set environment variable ``export SKIP_KUBE=1``.
+
+Additionally, all code must pass a suite of go linters. There is a pre-commit yaml file in the receptor repository that points to the linter suite. It is best practice to install the pre-commit yaml so that the linters run locally on each commit.
```
Comment (beeankha) [inline docs/read-the-docs/source/developer_guide.rst]: Capitalize "Go"

Comment (fosterseth) [reply docs/read-the-docs/source/developer_guide.rst]: fixed, and I changed all golang to Golang where applicable

File: docs/read-the-docs/source/developer_guide.rst (hunk 27 above)
Comment (fosterseth) [reply docs/read-the-docs/source/developer_guide.rst]: good suggestion, thanks

**Comment by fosterseth** [comment]
@thenets This PR moves the stuff you added in docs/ to tools/. That seems like a more proper place, similar to the tools directory in awx

**Comment by thenets** [comment]
@fosterseth cool :) looks a better place with the new docs been added.

**Review by shanemcd — APPROVED** [review]

**Pull request merged**